src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+            ImmutableArray<string>.Builder removedDesignTimeInputs = ImmutableArray.CreateBuilder<string>();
+            ImmutableArray<string>.Builder addedDesignTimeSharedInputs = ImmutableArray.CreateBuilder<string>();
+
+            foreach (string item in changes.Difference.AddedItems)
+            {
+                PreprocessAddItem(item);
+            }
+
+            foreach (string item in changes.Difference.RemovedItems)
+            {
+                // the RemovedItems doesn't have any project info any more so we can't tell if an item was a design time input or not
+                // and we can't check in AppliedValue because there could be an unapplied value that adds the item that is still in the queue
+                // so we have to just process the removed items in ApplyAsync and take advantage of the fact that its
+                // quick to remove from immutable collections, even if the item doesn't exist, because they're implemented as trees
+                removedDesignTimeInputs.Add(item);
+            }
+
+            foreach ((string oldName, string newName) in changes.Difference.RenamedItems)
+            {
+                // A rename is just an add and a remove
+                PreprocessAddItem(newName);
+                removedDesignTimeInputs.Add(oldName);
+            }
+
+            foreach (string item in changes.Difference.ChangedItems)
+            {
+                // Items appear in this set if an items metadata changes. 
+                // For example that could be when DesignTime goes from false to true, so we care about it, but we don't know which metadata changed or what the old value was
+                // so there is a high chance this is an irrelevant change to us.
+
+                // The actual logic here is deceptively simple, so what is actually happening is this:
+                // Previous Design Time   | New Design Time   | Action
+                // ----------------------------------------------------------------
+                // False                  | False             | PreprocessAddItem will return false, ApplyAsync will process remove, which is fast enough to not worry about
+                // False                  | True              | PreprocessAddItem will return true, ApplyAsync will process as an add, so all good
+                // True                   | False             | PreprocessAddItem will return false, ApplyAsync will process remove, so all good
+                // True                   | True              | PreprocessAddItem will return true, ApplyAsync will not process add because key already exists
+                if (!PreprocessAddItem(item))
+                {
+                    removedDesignTimeInputs.Add(item);
+                }
+            }
+
+            string rootNamespace = project.GetPropertyValue(ConfigurationGeneral.RootNamespaceProperty);
```

Comment:
This should be gathered by a "ConfigurationGeneral" rule (similar to your Compile rule) that is an input for HostBridge, here reading from the evaluated instance bypassing all the rule & caching layer.

Reply:
This has been resolved by changes due to the suggestion ProjectRuleSource hint in https://github.com/dotnet/project-system/pull/4349#discussion_r238940967